docs/rfcs/004-session-management.md

```diff
@@ -0,0 +1,125 @@
+
+Let's revisit the way we manage session state so that it becomes an Atom feature that users can understand and trust.
+
+## Motivation
+
+_pending_
+
+Problems to address:
+
+* The active project (which is deserialized on open, serialized on close) is chosen implicitly based on the set of project folders.
+* The active project isn't explicitly named anywhere in the UI. Two open Atom windows with two different projects do not obviously differ; two open Atom windows on the same project are not obviously the same.
+* Adding or removing a root folder silently changes the active project.
+* When the same project is open in two Atom windows at once, the last-closed window clobbers any project state saved by the first-closed.
+* Unsaved content in a TextBuffer is only accessible through the project that was open when it was created.
+
+## Explanation
+
+A _Session_ is the unit of persistent window state in Atom related to a coherent development effort. It includes:
+
+* The set of root folders;
+* Pane and dock configuration;
+* Open pane and dock items;
+* Cursor position, fold state, and other TextEditor characteristics;
+* Per-package [serialized state](https://flight-manual.atom.io/behind-atom/sections/serialization-in-atom/).
+
+Each Atom window is associated with one Session, and each Session is associated with at most one Atom window at a time.
+
+Each Session has a unique, user-customizable name.
+
+Other persistent state is tracked _per-file_, based on its inode or fileID to work through filesystem softlinks or hardlinks. This includes:
+
+* Unsaved buffer changes;
+* Per-file manually changed indentation, line ending, and encoding settings.
+
+### Add an explicit indication of the current Session to the tree-view
+
+At the top of the tree-view, display the current Project's name. Clicking it opens the project management UI described below.
+
+> TODO: graphic
+
+### Add a command-line argument to explicitly choose the Session
+
+When opening Atom from the command line with one or more locations specified:
+
+1. If exactly one non-open Session matches those paths, that Session is loaded into the newly opened Atom window.
+2. If no non-open Sessions match, a new Session is created for that Atom window.
+3. If multiple Sessions match, a new Session is created for that Atom window, and a notification is displayed allowing the user to switch to any of the matching Sessions and destroy the just-created one.
+
+In this context, "matches" means _every path in the set, normalized through realpath and case sensitivity, begins with the path of at least one project folder in the Session under consideration_, and, as a special case, _a Session containing zero project folders matches everything_.
```

Comment:
I'm curious if this is still too much magic. I'm trying to thread the needle between preventing the "oh shit, where did my work go?" reaction on every Atom load and not having too much implicit, poorly understood behavior.

Reply:
This makes sense to me